chemprop/uncertainty/uncertainty_calibrator.py

```diff
@@ -0,0 +1,680 @@
+        num_models: int,
+        dataset_type: str,
+        loss_function: str,
+        uncertainty_dropout_p: float,
+        dropout_sampling_size: int,
+        spectra_phase_mask: List[List[bool]],
+    ):
+        self.calibration_data = calibration_data
+        self.calibration_data_loader = calibration_data_loader
+        self.regression_calibrator_metric = regression_calibrator_metric
+        self.interval_percentile = interval_percentile
+        self.dataset_type = dataset_type
+        self.uncertainty_method = uncertainty_method
+        self.loss_function = loss_function
+        self.num_models = num_models
+
+        self.raise_argument_errors()
+
+        self.calibration_predictor = build_uncertainty_predictor(
+            test_data=calibration_data,
+            test_data_loader=calibration_data_loader,
+            models=models,
+            scalers=scalers,
+            num_models=num_models,
+            dataset_type=dataset_type,
+            loss_function=loss_function,
+            uncertainty_method=uncertainty_method,
+            uncertainty_dropout_p=uncertainty_dropout_p,
+            dropout_sampling_size=dropout_sampling_size,
+            individual_ensemble_predictions=False,
+            spectra_phase_mask=spectra_phase_mask,
+        )
+
+        self.calibrate()
+
+    @property
+    @abstractmethod
+    def label(self):
+        """
+        The string in saved results indicating the uncertainty method used.
+        """
+
+    def raise_argument_errors(self):
+        """
+        Raise errors for incompatibilities between dataset type and uncertainty method, or similar.
+        """
+        if self.dataset_type == "spectra":
+            raise NotImplementedError(
+                "No uncertainty calibrators are implemented for the spectra dataset type."
+            )
+        if self.uncertainty_method in ['ensemble', 'dropout'] and self.dataset_type in ['classification', 'multiclass']:
+            raise NotImplementedError(
+                'Though ensemble and dropout uncertainty methods are available for classification \
+                    multiclass dataset types, their outputs are not confidences and are not \
+                    compatible with any implemented calibration methods for classification.'
+            )
+
+    @abstractmethod
+    def calibrate(self):
+        """
+        Fit calibration method for the calibration data.
+        """
+
+    @abstractmethod
+    def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        """
+        Take in predictions and uncertainty parameters from a model and apply the calibration method using fitted parameters.
+        """
+
+    def nll(
+        self,
+        preds: List[List[float]],
+        unc: List[List[float]],
+        targets: List[List[float]],
+    ):
+        """
+        Takes in calibrated predictions and uncertainty parameters and returns the log probability density of that result.
+        """
```

Comment:
It's only the regression methods that need nll. Classification and multiclass have a consistent definitions that can be covered in the evaluator class instead of the specific calibrators. I could give the classification/multiclass methods a dummy nll to prevent the error. I'm torn by it being unnecessary for those and the need to force it for regression calibrators.

Reply:
Eh. Probably better as an abstract method so it's there for the class even if people want to do something other than the built in evaluator method with it.